### ISSUE TITLE
SIGSEGV after (lookup) in ban_evaluate

### ISSUE THREAD

**Issue by b3k** [varnishcache/varnish-cache#2214]
Our application cache REST API responses, and also invalidate certain cache objects using Varnish Admin interface. When warming-up cache for some requests everything is fine, problem occurs when we send invalidation request to Varnish Admin like:

`ban obj.http.X-Varnish-Tags ~ .*7b389\\(([A-Za-z0-9-_]+,)*a7a7e(,[A-Za-z0-9-_]+)*\\);.*`

and when we request object affected by ban we are receiving `502 Bad Gateway`.

After some investigation of the issue we have found that SIGSEGV starts just after `return (lookup)` in `vcl_hash` when matching regexp from ban against current request by ban lurker.

Stack dump (full: https://gist.github.com/b3k/3edc3e82477286e48aac3fc4a1cbbdeb)
Fragment:
```
   tables = <optimised out>
         start_bits = <optimised out>
         start_match = 0x7f0b9a01d0ab "2f913(4,5,8,a,b,c,d,f,10,11,12,13,14,15,16,18,19,1b,1c,1d,1e,20,21,23,25,2a,2c,2d,31,33,34,36,38,3a,3e,3f,43,45,47,49,4b,51,53,54,55,57,59,5c,5f,60,63,65,67,68,69,6b$
         end_subject = <optimised out>
         start_partial = 0x0
         req_char_ptr = 0x7f0b9a01d17b ";"
         study = 0x0
         re = 0x7f0bac640dec
 #117 0x0000000000418298 in ban_evaluate (wrk=wrk@entry=0x7f0ba1794c40, bs=<optimised out>, oc=oc@entry=0x7f0b9a441c40, reqhttp=0x7f0b9b496298, tests=tests@entry=0x7f0ba17941ac) at cache/cache_ban.c:475
         bt = {oper = 18 '\022', arg1 = 26 '\032', arg1_spec = 0x7f0bac640d91 "\017X-Varnish-Tags:", arg2 = 0x7f0bac640dac ".*2f913\\(([A-Za-z0-9-_]+,)*5dcb(,[A-Za-z0-9-_]+)*\\);.*", arg2_spec = 0x7f0bac6$
         be = 0x7f0bac640e99 ""
         p = 0x1dbfb2635ee72fff <error: Cannot access memory at address 0x1dbfb2635ee72fff>
         arg1 = 0x7f0b9a01d0ab "2f913(4,5,8,a,b,c,d,f,10,11,12,13,14,15,16,18,19,1b,1c,1d,1e,20,21,23,25,2a,2c,2d,31,33,34,36,38,3a,3e,3f,43,45,47,49,4b,51,53,54,55,57,59,5c,5f,60,63,65,67,68,69,6b,6c,6f,$
         __func__ = "ban_evaluate"
 #118 0x0000000000418433 in BAN_CheckObject (wrk=wrk@entry=0x7f0ba1794c40, oc=oc@entry=0x7f0b9a441c40, req=req@entry=0x7f0b9b496020) at cache/cache_ban.c:548
         b = 0x7f0bac642100
         vsl = 0x7f0b9b496258
         b0 = 0x7f0bac642bf0
         bn = 0x7f0bac641600
         tests = 20
         __func__ = "BAN_CheckObject"
 #119 0x0000000000429afd in HSH_Lookup (req=req@entry=0x7f0b9b496020, ocp=ocp@entry=0x7f0ba1794290, bocp=bocp@entry=0x7f0ba1794298, wait_for_busy=1, always_insert=0) at cache/cache_hash.c:424
         wrk = 0x7f0ba1794c40
         oh = 0x7f0b9a43de20
         oc = 0x7f0b9a441c40
         exp_oc = 0x0
         exp_t_origin = 0
         busy_found = 0
         retval = <optimised out>
         vary = <optimised out>
         __func__ = "HSH_Lookup"
 #120 0x00000000004370a3 in cnt_lookup (req=0x7f0b9b496020, wrk=0x7f0ba1794c40) at cache/cache_req_fsm.c:369
         oc = 0x0
         lr = <optimised out>
         had_objhead = 0
         boc = 0x0
```
varnishlog dosent give much:
```
*   << Session  >> 131777
-   Begin          sess 0 HTTP/1
-   SessOpen       10.1.0.20 33295 :80 10.1.0.230 80 1487147783.929461 23
-   Link           req 131778 rxreq
-   VSL            flush
-   End            synth

Log abandoned
Log reacquired
```

Syslog:
```
Feb 15 09:40:29 HyperionVarnish varnishd[10723]: Child (32753) not responding to CLI, killed it.
Feb 15 09:40:29 HyperionVarnish varnishd[10723]: CLI telnet 10.1.0.232 40212 10.1.0.230 6082 Wr 400 CLI communication error (hdr)
Feb 15 09:40:29 HyperionVarnish varnishd[10723]: Child (32753) died signal=11 (core dumped)
Feb 15 09:40:29 HyperionVarnish varnishd[10723]: Child cleanup complete
Feb 15 09:40:29 HyperionVarnish varnishd[10723]: Child (922) Started
```

## Expected Behavior
Request returns HTTP 200 OK with invalidated content.

## Current Behavior
Request returns HTTP 502 Bad Gateway.

## Steps to Reproduce
1. Request page that gives HTTP header with many (at least 80) items in format: `name(elem1,elem2,elem3,...);`. For example:
```
X-Varnish-Tags: 
2f913(4,5,7,8,a,b,c,d,f,10,11,12,13,14,15,16,18,19,1b,1c,1d,1e,20,21,23,25,27,2a,2c,2d,31,33,34,36,38,39,3a,3e,3f,43,45,47,49,4b,51,53,54,55,57,59,5c,5f,60,63,65,67,68,69,6b,6c,6f,71,73,75,76,78,7a,7c,7d,7e,80,83,85,87,89,8b,8c,8d,8f,91,93,95,97,98,99,9a,9c,9e,a0,a1,a2,a4,a6,a7,a8,a9,aa,ab,ac,ad);
```
2. Send ban using admin interface:
```
ban obj.http.X-Varnish-Tags ~ .*2f913\\(([A-Za-z0-9-_]+,)*2459(,[A-Za-z0-9-_]+)*\\);.*
```
3. Request this page second time to retrieve it from cache.

## Context
We have application that serves quite important REST API. We have implemented more reboost invalidation process for that API, and as a solution we are using popular method with tags in HTTP headers. It works well, but troubles begins when dealing with bigger collections.

## Environment
varnishd (varnish-4.1.5 revision 2c82b1c) (Varnish 5 is also affected)
Ubuntu 14.04.5 LTS
Linux HyperionVarnish 3.13.0-96-generic #143-Ubuntu SMP Mon Aug 29 20:15:20 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

**Comment by nigoroll** [comment]
your regexp is inefficient and you are running into massive recursions in pcre.
Increasing thread_pool_stack could help as an interims release, but the regexp should be fixed.
This is not a support channel, but some hints:
* no leading/trailing {{.*}}
* why the two {{([A-Za-z0-9-_]+,)*}} if all you want to do is match for {{[(,]2459[),]}} ?

**Comment by maciejmrozinski** [comment]
@nigoroll 
I must say I'm very negatively surprised by a form of Your reply... Here You have an examples that this regex works with PCRE:
1. https://regex101.com/r/Ov99A8/1
2. http://www.phpliveregex.com/p/j3K
3. http://sandbox.onlinephpfunctions.com/code/88451c93f3b3d4381fc2fadfb601bbfa2907fbfa

It may be not efficient, but still, Varnish is killed with segmentation fault, which is not good. This is an ISSUE and I cannot understand why You marked it as INVALID one. Varnish should gracefully log a problem IMO and not restart it's child. If i found 3 places where all this works and one place where it isn't - yes, Varnish is this one, so what do You think, where is the problem?

**Comment by gquintard** [comment]
The question is really not whether it works or not, what Nils says is that it's inefficient, and that you didn't provided enough stack space for PCRE to do its job properly.

It's akin to running any program without sufficient RAM, you will get OOM killed.

If you don't agree, please move this discussion to varnish-misc, or IRC (http://www.varnish-cache.org/support/) as we try to keep github issues for actionnable problems.

**Comment by nigoroll** [comment]
@maciejmrozinski Yes my quick response was lacking a bit of an explanation:
The regexp you're using forces pcre to recurse to a depth for which the stack space is not sufficient, the segmentation violation you are seeing is a stack overflow. We had discussed ideas about preventing this form of stack overflow in the past and to cut a long story short, there is no portable way to do so. And even if we prevented the actual stack overflow, the regexp would still fail with the amount of stack space available, so we would not know if the ban applied or not, so we would probably need to ban the object for all cases, which would, again, lead to more complications...
In short: The fix is to choose an efficient regexp, and I guess I've already given you the solution :)

**Comment by slimhazard** [comment]
Sorry, I can't help jumping in on this, it's a hobby horse of mine.

@maciejmrozinski, those online regex helpers are probably configured with very large stacks -- they'd have to be, in order to prevent "regex DOS" attacks due to the worst-case exponential running time of a regex engine with backtracking. And they do in fact cut off a regex match when backtracking and recursion become excessive. Have a look at this:

https://regex101.com/r/zJDYtC/1

Notice "catastrophic backtracking" in the red box. It took a much larger test string to do that, probably because regex101 runs the matcher with a much larger stack than your Varnish process.

Also, your online regex examples are all successful matches, but your crash happened on a non-match. It's the non-matches that can cause deep backtracking, since the matcher keeps searching until all possibilities are exhausted.

Take a look at: http://www.pcre.org/original/doc/html/pcrestack.html

Any program that makes an external library call that causes stack overflow _will_ crash with a segfault, with no way of recovering gracefully, regardless of the program or the library. The same thing would happen when the PCRE matcher is called from PHP, if recursion is deep enough and the stack is not large enough. So it's not a Varnish issue, because there's nothing that Varnish or any other client of PCRE could do about it.

About your regex: your crash dump shows that the `match` call in `pcre_exec` goes to recursion depth 115 when it runs out of stack, while it's matching `([A-Za-z0-9-_]+,)*` after the first parenthesis. That expression has nested quantifiers -- the `+` quantifier within an expression with the `*` quantifier. Those are very common culprits for deep recursion and stack overflow.

To restate what @nigoroll has said: Does your ban _have_ to have that kind of "validation" of the hex data before and after the part of the string that evidently really matters, in this case `2459`, with or without commas before and after?

How about: `\b2f913\([^)]*,?2459(?:,|\))`
(You may have to escape the backslashes as double backslashes.)

"After a word boundary, the string `2f913(` followed by zero or more characters that are not `)`, then possibly a comma, followed by `2459`,  followed by either a comma or `)`."

That works: https://regex101.com/r/FAKSSm/1

**Comment by hermunn** [comment]
Backport review: Nothing commited, nothing to backport.
